dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/cdi/DataExtension.java

```diff
@@ -439,8 +440,15 @@ private boolean discoverEntityClasses(AnnotatedType<?> repositoryType,
                 } else if (c.isArray()) {
                     c = c.getComponentType();
                 }
-                if (Object.class.equals(c)) { // generic parameter like BasicRepository.save(S entity)
-                    entityParamType = method.getParameterTypes()[0];
+                if (Object.class.equals(c)) {
+                    // generic parameter like BasicRepository.save(S entity) or BasicRepository.deleteById(@By(ID) K id)
+                    boolean isEntity = true;
+                    for (Annotation anno : method.getParameterAnnotations()[0])
+                        if (By.class.equals(anno.annotationType()))
+                            isEntity = false;
+                    if (isEntity)
+                        entityParamType = c;
+                    // TODO is there any way to distinguish @Delete deleteById(K key) when @By is not present?
```

Comment:
IIRC, for Jakarta Persistence the Id of an entity must be a Java primitive or a primitive wrapper. So maybe something like, 

```java
for (Annotation anno : method.getParameterAnnotations()[0])
    if (By.class.equals(anno.annotationType()))
        isEntity = false;
if (method.getParameters[0].getClass().isPrimitive())
  isEntity=false;
if(isPrimitiveWrapper(method.getParameters[0].getClass()) //custom helper method
  isEntity=false;
if (isEntity)
    entityParamType = c;
```

Though maybe this needs to be clarified in the spec since it might not be generally true for other entities.

Reply:
UUID and String are also a valid types of Id in Jakarta Persistence, but yes, that seems like a good solution.